src/mcts/search.cc

```diff
@@ -1466,7 +1441,88 @@ void SearchWorker::DoBackupUpdateSingleNode(
   search_->total_playouts_ += node_to_process.multivisit;
   search_->cum_depth_ += node_to_process.depth * node_to_process.multivisit;
   search_->max_depth_ = std::max(search_->max_depth_, node_to_process.depth);
-}  // namespace lczero
+}
+
+bool SearchWorker::MaybeConvertParent(Node* p, float m) const {
+  auto found_tb = false;
+  auto losing_m = 0.0f;
+
+  // Determine the maximum (Lower, Upper) bounds across all children.
+  // (-1,-1) Loss (initial and lowest bounds)
+  // (-1, 0) Can't Win
+  // (-1, 1) Regular node
+  // ( 0, 0) Draw
+  // ( 0, 1) Can't Lose
+  // ( 1, 1) Win (highest bounds)
+  auto max_L = -1;
+  auto max_U = -1;
+  for (const auto& edge : p->Edges()) {
+    if (edge.IsTerminal()) {
+      found_tb = found_tb || edge.IsTbTerminal();
+
+      // Default_wl doesn't matter as WL is only used if IsTerminal is true.
+      const auto WL = edge.GetWL(0.0f);
+      // We can short-circuit after a single win maxes the bounds.
+      if (WL > 0.0f) {
+        max_L = max_U = 1;
+        break;
+      } else if (WL == 0.0f) {
+        max_L = std::max(0, max_L);
+        max_U = std::max(0, max_U);
+      } else {
+        // Track the longest loss.
+        losing_m = std::max(losing_m, edge.GetM(0.0f));
+      }
+    } else if (edge.CantWin()) {
+      max_U = std::max(0, max_U);
+    } else if (edge.CantLose()) {
+      max_L = std::max(0, max_L);
+      max_U = 1;
+    } else {
+      // A regular or unvisited node maxes the upper bound.
+      max_U = 1;
+    }
+  }
+
+  auto parent_type =
+      found_tb ? Node::NodeType::TablebaseTerminal : Node::NodeType::Terminal;
+
+  // The parent's bounds are flipped from the children (-max(U), -max(L))
+  // aggregated as if it was a single child (forced move) of the same bound.
+  //       Loss (-1,-1) -> ( 1, 1) Win
+  //  Can't Win (-1, 0) -> ( 0, 1) Can't Lose
+  //    Regular (-1, 1) -> (-1, 1) Regular
+  //       Draw ( 0, 0) -> ( 0, 0) Draw
+  // Can't Lose ( 0, 1) -> (-1, 0) Can't Win
+  //        Win ( 1, 1) -> (-1,-1) Loss
+  if (max_L == -1) {
+    if (max_U == -1) {
+      p->MakeTerminal(GameResult::WHITE_WON, std::max(losing_m, m) + 1.0f,
+                      parent_type);
+    } else if (max_U == 0) {
+      // No moves can win, so remove losing from parent.
+      p->RemoveOutcome(GameResult::BLACK_WON);
```

Comment:
It feels like `SetOutcome` or even `SetNodeType` would be more readable.

Reply:
I guess part of the reason why it's hard to read is from reusing `GameResult`

Reply:
Changed to `p->SetBounds(…)`